### ISSUE TITLE
consulCatalog provider fails: too many colons in address

### ISSUE THREAD

**Issue by ibayer** [traefik/traefik#7428]
### Do you want to request a *feature* or report a *bug*?

Bug, that renders consulCatalog provider unusable.

### What did you do?

I'm testing if traefik v2 works with consulCatalog provider. I didn't experience this issue in `v1.7.3`.

### What did you expect to see?

No errors.

### What did you see instead?

```bash
time="2020-10-16T10:20:28Z" level=error msg="Provider connection error error create consul client, address 

http://127.0.0.1:8500: too many colons in address, retrying in 552.330144ms" providerName=consulcatalog
```

### Output of `traefik version`: (_What version of Traefik are you using?_)

Same issue with the following images:

- image = `traefik:v2.2.0`
- mage = `traefik:v2.3.0`
- image = `traefik:v2.3.1` (provided logs are from this tag)

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

```yml
entryPoints:
  http:
    address: :80
  https:
    address: :443
  traefik:
    address: :8081

http:
  middlewares:
    # Redirect to https
    test-redirectscheme:
      redirectScheme:
        scheme: https
        permanent: true

  routers:
    ## Dynamic configuration
    routerfoo:
      tls:
          certResolver: default

api:
  insecure: true
  dashboard: true

log:
  level: DEBUG

certificatesResolvers:
  default:
    acme:
      email: admin@xxx.io
      storage: acme.json

# Default base domain used for the frontend rules.
# domain = "xxx.eu"
# domain = "xxx.io"

# ##############################################################
# Consul Catalog Provider
# ##############################################################

providers:
  providersThrottleDuration: 2s
  consulCatalog:
    prefix: traefik
    exposedByDefault: false
    stale: true
    endpoint:
      # Defines the Consul server endpoint.
      # address: "https://127.0.0.1:8500"
      tls:
        ca: /etc/consul.d/ca.pem
        cert: /etc/consul.d/cli.pem
        key: /etc/consul.d/cli-key.pem
```

Note: I see the same error independent if `address: "https://127.0.0.1:8500"` is set explicitly or not.

### If applicable, please paste the log output in DEBUG level (`--log.level=DEBUG` switch)

```bash
dpoint\":{\"address\":\"http://127.0.0.1:8500\",\"tls\":{\"ca\":\"/etc/consul.d/ca.pem\",\"cert\":\"/etc/consul.d/cli.pem\",\"key\":\"/etc/consul.d/cli-key.pem\"}},\"prefix\":\"traefik\",\"refreshInterval\":15000000000,\"stale\":true,\"defaultRule\":\"Host(`{{ normalize .Name }}`)\"}"
time="2020-10-16T10:39:51Z" level=error msg="Provider connection error error create consul client, address http://127.0.0.1:8500: too many colons in address, retrying in 552.330144ms" providerName=consulcatalog
time="2020-10-16T10:39:51Z" level=debug msg="Configuration received from provider default.acme: {\"http\":{},\"tls\":{}}" providerName=default.acme
time="2020-10-16T10:39:51Z" level=debug msg="Configuration received from provider internal: {\"http\":{\"routers\":{\"api\":{\"entryPoints\":[\"traefik\"],\"service\":\"api@internal\",\"rule\":\"PathPrefix(`/api`)\",\"priority\":2147483646},\"dashboard\":{\"entryPoints\":[\"traefik\"],\"middlewares\":[\"dashboard_redirect@internal\",\"dashboard_stripprefix@internal\"],\"service\":\"dashboard@internal\",\"rule\":\"PathPrefix(`/`)\",\"priority\":2147483645}},\"services\":{\"api\":{},\"dashboard\":{},\"noop\":{}},\"middlewares\":{\"dashboard_redirect\":{\"redirectRegex\":{\"regex\":\"^(http:\\\\/\\\\/(\\\\[[\\\\w:.]+\\\\]|[\\\\w\\\\._-]+)(:\\\\d+)?)\\\\/$\",\"replacement\":\"${1}/dashboard/\",\"permanent\":true}},\"dashboard_stripprefix\":{\"stripPrefix\":{\"prefixes\":[\"/dashboard/\",\"/dashboard\"]}}}},\"tcp\":{},\"tls\":{}}" providerName=internal
time="2020-10-16T10:39:51Z" level=debug msg="Adding certificate for domain(s) count.xxx.io"
time="2020-10-16T10:39:51Z" level=debug msg="No default certificate, generating one"
time="2020-10-16T10:39:51Z" level=error msg="Provider connection error error create consul client, address http://127.0.0.1:8500: too many colons in address, retrying in 1.080381816s" providerName=consulcatalog
time="2020-10-16T10:39:51Z" level=debug msg="Added outgoing tracing middleware api@internal" routerName=api@internal middlewareName=tracing middlewareType=TracingForwarder entryPointName=traefik
time="2020-10-16T10:39:51Z" level=debug msg="Added outgoing tracing middleware dashboard@internal" routerName=dashboard@internal middlewareName=tracing middlewareType=TracingForwarder entryPointName=traefik
time="2020-10-16T10:39:51Z" level=debug msg="Creating middleware" middlewareType=StripPrefix entryPointName=traefik routerName=dashboard@internal middlewareName=dashboard_stripprefix@internal
time="2020-10-16T10:39:51Z" level=debug msg="Adding tracing to middleware" entryPointName=traefik routerName=dashboard@internal middlewareName=dashboard_stripprefix@internal
time="2020-10-16T10:39:51Z" level=debug msg="Creating middleware" routerName=dashboard@internal entryPointName=traefik middlewareName=dashboard_redirect@internal middlewareType=RedirectRegex
time="2020-10-16T10:39:51Z" level=debug msg="Setting up redirection from ^(http:\\/\\/(\\[[\\w:.]+\\]|[\\w\\._-]+)(:\\d+)?)\\/$ to ${1}/dashboard/" middlewareType=RedirectRegex routerName=dashboard@internal entryPointName=traefik middlewareName=dashboard_redirect@internal
time="2020-10-16T10:39:51Z" level=debug msg="Adding tracing to middleware" entryPointName=traefik routerName=dashboard@internal middlewareName=dashboard_redirect@internal
time="2020-10-16T10:39:51Z" level=debug msg="Creating middleware" middlewareName=traefik-internal-recovery middlewareType=Recovery entryPointName=traefik
time="2020-10-16T10:39:52Z" level=error msg="Provider connection error error create consul client, address http://127.0.0.1:8500: too many colons in address, retrying in 1.31013006s" providerName=consulcatalog
time="2020-10-16T10:39:54Z" level=error msg="Provider connection error error create consul client, address http://127.0.0.1:8500: too many colons in address, retrying in 1.582392691s" providerName=consulcatalog

```

**Comment by ldez** [comment]
Hello,

could you try `address: "127.0.0.1:8500"`?

Thank you

**Comment by ibayer** [comment]
Thanks, this does solve the original error.
I'm now seeing a new one

```bash
time="2020-10-16T14:03:10Z" level=error msg="error get consul catalog data, Unexpected response code: 400 (Client sent an HTTP request to an HTTPS server.\n)" providerName=consulcatalog
```

log:

```bash
time="2020-10-16T14:02:55Z" level=debug msg="Adding certificate for domain(s) count.palaimon.io"
time="2020-10-16T14:02:55Z" level=debug msg="No default certificate, generating one"
time="2020-10-16T14:02:55Z" level=info msg="Error renewing certificate from LE : {Main:youtrack.palaimon.io SANs:[]}, ACME challenge not specified, please select TLS or HTTP or DNS Challenge" providerName=default.acme
time="2020-10-16T14:02:55Z" level=debug msg="Building ACME client..." providerName=default.acme
time="2020-10-16T14:02:55Z" level=debug msg="https://acme-v02.api.letsencrypt.org/directory" providerName=default.acme
time="2020-10-16T14:02:56Z" level=debug msg="Added outgoing tracing middleware dashboard@internal" entryPointName=traefik routerName=dashboard@internal middlewareName=tracing middlewareType=TracingForwarder
time="2020-10-16T14:02:56Z" level=debug msg="Creating middleware" entryPointName=traefik routerName=dashboard@internal middlewareName=dashboard_stripprefix@internal middlewareType=StripPrefix
time="2020-10-16T14:02:56Z" level=debug msg="Adding tracing to middleware" entryPointName=traefik routerName=dashboard@internal middlewareName=dashboard_stripprefix@internal
time="2020-10-16T14:02:56Z" level=debug msg="Creating middleware" entryPointName=traefik routerName=dashboard@internal middlewareName=dashboard_redirect@internal middlewareType=RedirectRegex
time="2020-10-16T14:02:56Z" level=debug msg="Setting up redirection from ^(http:\\/\\/(\\[[\\w:.]+\\]|[\\w\\._-]+)(:\\d+)?)\\/$ to ${1}/dashboard/" entryPointName=traefik routerName=dashboard@internal middlewareName=dashboard_redirect@internal middlewareType=RedirectRegex
time="2020-10-16T14:02:56Z" level=debug msg="Adding tracing to middleware" middlewareName=dashboard_redirect@internal entryPointName=traefik routerName=dashboard@internal
time="2020-10-16T14:02:56Z" level=debug msg="Added outgoing tracing middleware api@internal" middlewareType=TracingForwarder entryPointName=traefik routerName=api@internal middlewareName=tracing
time="2020-10-16T14:02:56Z" level=debug msg="Creating middleware" middlewareName=traefik-internal-recovery middlewareType=Recovery entryPointName=traefik
time="2020-10-16T14:02:56Z" level=info msg="Error renewing certificate from LE : {Main:palaimon.io SANs:[www.palaimon.io]}, ACME challenge not specified, please select TLS or HTTP or DNS Challenge" providerName=default.acme
time="2020-10-16T14:02:56Z" level=debug msg="Building ACME client..." providerName=default.acme
time="2020-10-16T14:02:56Z" level=debug msg="https://acme-v02.api.letsencrypt.org/directory" providerName=default.acme
time="2020-10-16T14:02:57Z" level=info msg="Error renewing certificate from LE : {Main:gitlab.palaimon.io SANs:[]}, ACME challenge not specified, please select TLS or HTTP or DNS Challenge" providerName=default.acme
time="2020-10-16T14:03:10Z" level=error msg="error get consul catalog data, Unexpected response code: 400 (Client sent an HTTP request to an HTTPS server.\n)" providerName=consulcatalog
time="2020-10-16T14:03:10Z" level=error msg="Provider connection error Unexpected response code: 400 (Client sent an HTTP request to an HTTPS server.\n), retrying in 357.131936ms" providerName=consulcatalog
time="2020-10-16T14:03:25Z" level=error msg="error get consul catalog data, Unexpected response code: 400 (Client sent an HTTP request to an HTTPS server.\n)" providerName=consulcatalog
time="2020-10-16T14:03:25Z" level=error msg="Provider connection error Unexpected response code: 400 (Client sent an HTTP request to an HTTPS server.\n), retrying in 440.328595ms" providerName=consulcatalog
```

**Comment by ibayer** [comment]
```bash
time="2020-10-16T14:03:10Z" level=error msg="error get consul catalog data, Unexpected response code: 400 (Client sent an HTTP request to an HTTPS server.\n)" providerName=consulcatalog
```

fixed by adding

```
providers:
  consulCatalog:
    endpoint:
      scheme: https
```

to config

**Comment by ibayer** [comment]
I can't get the lets-encrypt setup to work with the consule catalog provider. Looking at the debug log it looks to me like the colon issue `78.49.99.163:57812:` might exist at other places as well.

```bash
time="2020-10-16T18:16:10Z" level=debug msg="http: panic serving 78.49.99.163:57812: runtime error:
 invalid memory address or nil pointer dereference"
```

`acme.json` is empyt

log
```bash
time="2020-10-16T18:16:10Z" level=info msg="Configuration loaded from file: /traefik.yml"
time="2020-10-16T18:16:10Z" level=info msg="Traefik version 2.3.1 built on 2020-09-29T15:49:06Z"
time="2020-10-16T18:16:10Z" level=debug msg="Static configuration loaded {\"global\":{},\"serversTransport\":{\"maxIdleConnsPerHost\":200},\"entryPoints\":{\"traefik\":{\"address\":\":8083\",\"transport\":{\"lifeCycle\":{\"graceTimeOut\":10000000000},\"respondingTimeouts\":{\"idleTimeout\":180000000000}},\"forwardedHeaders\":{},\"http\":{}},\"web\":{\"address\":\":80\",\"transport\":{\"lifeCycle\":{\"graceTimeOut\":10000000000},\"respondingTimeouts\":{\"idleTimeout\":180000000000}},\"forwardedHeaders\":{},\"http\":{}},\"web-secure\":{\"address\":\":443\",\"transport\":{\"lifeCycle\":{\"graceTimeOut\":10000000000},\"respondingTimeouts\":{\"idleTimeout\":180000000000}},\"forwardedHeaders\":{},\"http\":{}}},\"providers\":{\"providersThrottleDuration\":2000000000,\"consulCatalog\":{\"endpoint\":{\"address\":\"127.0.0.1:8500\",\"scheme\":\"https\",\"tls\":{\"ca\":\"/etc/consul.d/ca.pem\",\"cert\":\"/etc/consul.d/cli.pem\",\"key\":\"/etc/consul.d/cli-key.pem\"}},\"prefix\":\"traefik2\",\"refreshInterval\":15000000000,\"stale\":true,\"defaultRule\":\"Host(`{{ normalize .Name }}`)\"}},\"api\":{\"insecure\":true,\"dashboard\":true},\"log\":{\"level\":\"DEBUG\",\"format\":\"common\"},\"certificatesResolvers\":{\"letsencrypt\":{\"acme\":{\"email\":\"admin@xxx.io\",\"caServer\":\"https://acme-staging-v02.api.letsencrypt.org/directory\",\"storage\":\"acme.json\",\"keyType\":\"RSA4096\",\"tlsChallenge\":{}}}}}"
time="2020-10-16T18:16:10Z" level=info msg="\nStats collection is disabled.\nHelp us improve Traefik by turning this feature on :)\nMore details on: https://doc.traefik.io/traefik/contributing/data-collection/\n"
time="2020-10-16T18:16:10Z" level=info msg="Starting provider aggregator.ProviderAggregator {}"
time="2020-10-16T18:16:10Z" level=debug msg="Start TCP Server" entryPointName=traefik
time="2020-10-16T18:16:10Z" level=debug msg="Start TCP Server" entryPointName=web
time="2020-10-16T18:16:10Z" level=debug msg="Start TCP Server" entryPointName=web-secure
time="2020-10-16T18:16:10Z" level=info msg="Starting provider *acme.Provider {\"email\":\"admin@xxx.io\",\"caServer\":\"https://acme-staging-v02.api.letsencrypt.org/directory\",\"storage\":\"acme.json\",\"keyType\":\"RSA4096\",\"tlsChallenge\":{},\"ResolverName\":\"letsencrypt\",\"store\":{},\"ChallengeStore\":{}}"
time="2020-10-16T18:16:10Z" level=info msg="Testing certificate renew..." providerName=letsencrypt.acme
time="2020-10-16T18:16:10Z" level=info msg="Starting provider *consulcatalog.Provider {\"endpoint\":{\"address\":\"127.0.0.1:8500\",\"scheme\":\"https\",\"tls\":{\"ca\":\"/etc/consul.d/ca.pem\",\"cert\":\"/etc/consul.d/cli.pem\",\"key\":\"/etc/consul.d/cli-key.pem\"}},\"prefix\":\"traefik2\",\"refreshInterval\":15000000000,\"stale\":true,\"defaultRule\":\"Host(`{{ normalize .Name }}`)\"}"
time="2020-10-16T18:16:10Z" level=info msg="Starting provider *traefik.Provider {}"
time="2020-10-16T18:16:10Z" level=debug msg="Configuration received from provider letsencrypt.acme: {\"http\":{},\"tls\":{}}" providerName=letsencrypt.acme
time="2020-10-16T18:16:10Z" level=debug msg="No default certificate, generating one"
time="2020-10-16T18:16:10Z" level=debug msg="Configuration received from provider internal: {\"http\":{\"routers\":{\"api\":{\"entryPoints\":[\"traefik\"],\"service\":\"api@internal\",\"rule\":\"PathPrefix(`/api`)\",\"priority\":2147483646},\"dashboard\":{\"entryPoints\":[\"traefik\"],\"middlewares\":[\"dashboard_redirect@internal\",\"dashboard_stripprefix@internal\"],\"service\":\"dashboard@internal\",\"rule\":\"PathPrefix(`/`)\",\"priority\":2147483645}},\"services\":{\"api\":{},\"dashboard\":{},\"noop\":{}},\"middlewares\":{\"dashboard_redirect\":{\"redirectRegex\":{\"regex\":\"^(http:\\\\/\\\\/(\\\\[[\\\\w:.]+\\\\]|[\\\\w\\\\._-]+)(:\\\\d+)?)\\\\/$\",\"replacement\":\"${1}/dashboard/\",\"permanent\":true}},\"dashboard_stripprefix\":{\"stripPrefix\":{\"prefixes\":[\"/dashboard/\",\"/dashboard\"]}}}},\"tcp\":{},\"tls\":{}}" providerName=internal
time="2020-10-16T18:16:10Z" level=debug msg="http: panic serving 78.49.99.163:57812: runtime error: invalid memory address or nil pointer dereference"
time="2020-10-16T18:16:10Z" level=debug msg="goroutine 47 [running]:"
time="2020-10-16T18:16:10Z" level=debug msg="net/http.(*conn).serve.func1(0xc0000b2280)"
time="2020-10-16T18:16:10Z" level=debug msg="\t/usr/local/go/src/net/http/server.go:1801 +0x147"
time="2020-10-16T18:16:10Z" level=debug msg="panic(0x2eb7e40, 0x54ceee0)"
time="2020-10-16T18:16:10Z" level=debug msg="\t/usr/local/go/src/runtime/panic.go:975 +0x3e9"
time="2020-10-16T18:16:10Z" level=debug msg="crypto/tls.(*Conn).readClientHello(0xc00049f180, 0x7efc17938108, 0x0, 0xc0008ada28)"
time="2020-10-16T18:16:10Z" level=debug msg="\t/usr/local/go/src/crypto/tls/handshake_server.go:139 +0x93"
time="2020-10-16T18:16:10Z" level=debug msg="crypto/tls.(*Conn).serverHandshake(0xc00049f180, 0xc000029ba0, 0x12)"
time="2020-10-16T18:16:10Z" level=debug msg="\t/usr/local/go/src/crypto/tls/handshake_server.go:40 +0x45"
time="2020-10-16T18:16:10Z" level=debug msg="crypto/tls.(*Conn).Handshake(0xc00049f180, 0x0, 0x0)"
time="2020-10-16T18:16:10Z" level=debug msg="\t/usr/local/go/src/crypto/tls/conn.go:1362 +0xc9"
time="2020-10-16T18:16:10Z" level=debug msg="net/http.(*conn).serve(0xc0000b2280, 0x39eadc0, 0xc000432600)"
time="2020-10-16T18:16:10Z" level=debug msg="\t/usr/local/go/src/net/http/server.go:1817 +0x1a5"
time="2020-10-16T18:16:10Z" level=debug msg="created by net/http.(*Server).Serve"
time="2020-10-16T18:16:10Z" level=debug msg="\t/usr/local/go/src/net/http/server.go:2969 +0x36c"
time="2020-10-16T18:16:10Z" level=debug msg="Added outgoing tracing middleware dashboard@internal" middlewareType=TracingForwarder entryPointName=traefik routerName=dashboard@internal middlewareName=tracing
time="2020-10-16T18:16:10Z" level=debug msg="Creating middleware" entryPointName=traefik routerName=dashboard@internal middlewareName=dashboard_stripprefix@internal middlewareType=StripPrefix
time="2020-10-16T18:16:10Z" level=debug msg="Adding tracing to middleware" entryPointName=traefik routerName=dashboard@internal middlewareName=dashboard_stripprefix@internal
time="2020-10-16T18:16:10Z" level=debug msg="Creating middleware" routerName=dashboard@internal middlewareType=RedirectRegex middlewareName=dashboard_redirect@internal entryPointName=traefik
time="2020-10-16T18:16:10Z" level=debug msg="Setting up redirection from ^(http:\\/\\/(\\[[\\w:.]+\\]|[\\w\\._-]+)(:\\d+)?)\\/$ to ${1}/dashboard/" middlewareType=RedirectRegex middlewareName=dashboard_redirect@internal entryPointName=traefik routerName=dashboard@internal
time="2020-10-16T18:16:10Z" level=debug msg="Adding tracing to middleware" entryPointName=traefik routerName=dashboard@internal middlewareName=dashboard_redirect@internal
time="2020-10-16T18:16:10Z" level=debug msg="Added outgoing tracing middleware api@internal" middlewareName=tracing middlewareType=TracingForwarder entryPointName=traefik routerName=api@internal
time="2020-10-16T18:16:10Z" level=debug msg="Creating middleware" middlewareName=traefik-internal-recovery middlewareType=Recovery entryPointName=traefik
time="2020-10-16T18:16:10Z" level=debug msg="No default certificate, generating one"
time="2020-10-16T18:16:10Z" level=debug msg="Serving default certificate for request: \"nextcloud.xxx.io\""
time="2020-10-16T18:16:10Z" level=debug msg="Serving default certificate for request: \"youtrack.xxx.io\""
```

updated config

```
global:
  checkNewVersion: false
  sendAnonymousUsage: false

entryPoints:
  web:
    address: :80
  web-secure:
    address: :443
  traefik:
    address: :8083

http:
  routers:
      web-router:
        entryPoints:
          - web
        middlewares:
          - redirect

      web-secure-router:
        entryPoints:
          - web-secure
        tls:
          # https://docs.traefik.io/master/routing/routers/#certresolver
          certResolver: letsencrypt

  middlewares:
    redirect:
      # https://docs.traefik.io/master/middlewares/redirectscheme/
      redirectScheme:
        scheme: https
        permanent: true

api:
  insecure: true
  dashboard: true

log:
  level: DEBUG

certificatesResolvers:
  # https://docs.traefik.io/master/https/acme/#lets-encrypt
  letsencrypt:
    acme:
      email: admin@xxx.io
      storage: acme.json
      # https://docs.traefik.io/master/https/acme/#httpchallenge
      caServer: https://acme-staging-v02.api.letsencrypt.org/directory
     #httpChallenge:
     #  entryPoint: web
      tlsChallenge: {}

# ##############################################################
# Consul Catalog Provider
# ##############################################################
#

providers:
  providersThrottleDuration: 2s
  consulCatalog:
    prefix: traefik2
    exposedByDefault: false
    stale: true
    endpoint:
      # Defines the Consul server endpoint.
      scheme: https
      address: "127.0.0.1:8500"
      tls:
        ca: /etc/consul.d/ca.pem
        cert: /etc/consul.d/cli.pem
        key: /etc/consul.d/cli-key.pem
```

**Comment by ldez** [comment]
It's not a semicolon issue, the error message template (inside Go) contains the last semicolon:
```go
c.server.logf("http: panic serving %v: %v\n%s", c.remoteAddr, err, buf)
```

Do you get this error all the time or did you only get it once?

**Comment by ibayer** [comment]
Thanks for the help. The config below finally worked. :)

<details><summary>working config</summary>

```yml

global:
  checkNewVersion: false
  sendAnonymousUsage: false


entryPoints:
  web:
    address: :80
    http:
      redirections:
        entryPoint:
          to: web-secure
          scheme: https
          permanent: true

  web-secure:
    address: :443
  traefik:
    address: :8081

api:
  insecure: true
  dashboard: true

log:
  level: DEBUG

certificatesResolvers:
  # https://docs.traefik.io/master/https/acme/#lets-encrypt
  letsencrypt:
    acme:
      email: admin@xxx.io
      storage: acme.json
      # https://docs.traefik.io/master/https/acme/#httpchallenge
      # comment for production certs
      # caServer: https://acme-staging-v02.api.letsencrypt.org/directory
      caServer: https://acme-v02.api.letsencrypt.org/directory
      httpChallenge:
        entryPoint: web
     #tlsChallenge: {}


providers:
  # https://docs.traefik.io/master/providers/file/
  file:
    filename: /local/traefik.yml
    watch: true
  providersThrottleDuration: 2s
  consulCatalog:
    prefix: traefik2
    exposedByDefault: false
    stale: true
    endpoint:
      # Defines the Consul server endpoint.
      scheme: https
      address: "127.0.0.1:8500"
      tls:
        ca: /etc/consul.d/ca.pem
        cert: /etc/consul.d/cli.pem
        key: /etc/consul.d/cli-key.pem
```
</details>

**Comment by traefiker** [comment]
Closed by #7429.
